### DISCUSSION TITLE
why the field ps.n is a string?

### DISCUSSION THREAD

**Discussion by l1t1** [duckdb/duckdb#7474] [Q&A] [answered]
with recursive points(n,x,y)
as (select row_number()over(),x,y
from generate_series(1,13)t(x),
generate_series(1,3)t2(y)
),
lines(m,psn,pen,len,passmap)as
(select row_number()over(),ps.n,pe.n,abs(ps.x-pe.x)+abs(ps.y-pe.y),
case when ps.x=pe.x then 
(select bit_or(1<<t.i) from generate_series(ps.n::int,pe.n::int,sign(pe.y-ps.y)*13::int)t(i)where i<>pe.n)
when ps.y=pe.y then (select bit_or(1<<i) from generate_series(ps.n,pe.n,sign(pe.x-ps.x))t(i) where i<>pe.n)
end
from points ps,points pe 
where (ps.x=pe.x and ps.y<>pe.y) or (ps.x<>pe.x and ps.y=pe.y))
select * from lines;
Error: Conversion Error: Could not convert string 'ps.n' to INT32

**Comment by Tishj** [comment, marked as answer]
`generate_series` is not a table in-out function (yet)
It only accepts constant expressions currently, that's why `ps.n` is seen as a string literal

**Reply by l1t1** [reply]
thank you

**Comment by l1t1** [comment]
why these sql works?
```
D select generate_series(len(A)) from (select range(7,10,2)A);
┌───────────────────────────┐
│ generate_series(len("A")) │
│          int64[]          │
├───────────────────────────┤
│ [0, 1, 2]                 │
└───────────────────────────┘
D create table x as select 12 A;
D select generate_series(len(A)) from x;
┌─────────────────────────┐
│ generate_series(len(A)) │
│         int64[]         │
├─────────────────────────┤
│ [0, 1, 2]               │
└─────────────────────────┘
```

**Reply by l1t1** [reply]
`select generate_series(var)` is ok
`select sth from generate_series(var)` is wrong
so following sql works
```
with recursive points(n,x,y)
as (select row_number()over(),x,y
from generate_series(1,13)t(x),
generate_series(1,3)t2(y)
),
lines(m,psn,pen,len,passmap)as
(select row_number()over(),ps.n,pe.n,abs(ps.x-pe.x)+abs(ps.y-pe.y),
case when ps.x=pe.x then
(select generate_series(ps.n::int,pe.n::int,sign(pe.y-ps.y)*13::int))
when ps.y=pe.y then (select generate_series(ps.n,pe.n,sign(pe.x-ps.x)))
end
from points ps,points pe
where (ps.x=pe.x and ps.y<>pe.y) or (ps.x<>pe.x and ps.y=pe.y))
select * from lines;
```

**Reply by Tishj** [reply]
One is a scalar function, the other is a table function
`range | generate_series` is not a table in-out function yet, so it doesn't accept subqueries

#7522 is meant to fix this.
